### PR TITLE
feat(marshal): passable to quasi-quoted Justin expr

### PR DESCRIPTION
**Pull request by erights**
Closes: #XXXX
Refs: https://github.com/Agoric/agoric-sdk/pull/5401

## Description

We often need to see the deep contents of a Passable. Indeed, this was kicked off by a slack thread about wanting to quote
```js
M.splitRecord(
    harden({
      assetKind: M.or('nat', 'set', 'copySet', 'copyBag'),
      decimalPlaces: M.and(M.gte(-100), M.lte(100)),
    }),
  )
```

This PR provides a `qp` function meaning "quote passable". You can use it where you'd use `q`. It produces a quasi-quoted Justin expression of the form in which the passable would be passed. For example, the `qp` of the above pattern is

```js
`makeTagged("match:splitRecord",  [
  {
    assetKind: makeTagged("match:or", [
      "nat",
      "set",
      "copySet",
      "copyBag",
    ]),
    decimalPlaces: makeTagged("match:and", [
      makeTagged("match:gte", -100),
      makeTagged("match:lte", 100),
    ]),
  },
])`
```

### Security Considerations

Unlike `q`, `qp` does not lazily postpone rendering the expression into Justin. `q` can do this easily because it is inside the `'ses'` assert.js module and can use the rights amplification available only there. `qp` depends on `'@endo/marshal'` which is several levels up from `'ses'`. Rather than expose these rights amplification powers (or several other kludges that failed), I just made `qp` non-lazy. Thus there are no security implications.

### Scaling Considerations

Because `qp` is not lazy, simply constructing an error with `Fail` or `makeError(X...)` will pay all the cost of rendering to Justin, whether or not that error is ever used. 

### Documentation Considerations

Because Justin is a subset of HardenedJS, the Justin expressions produced by `qp` can be used in documentation if needed. Though see Compatibility Considerations below.

### Testing Considerations

- [ ] Only minimal tests written for `qp`. We should also write a test that loops through all the `jsonJustinPairs`, just as other Justin tests do.

### Compatibility Considerations

Using Justin with symbols exposes in the rendered string how Justin currently represents passable symbols. That changes in #2793. Whichever of these gets merged first, the other will have to reconcile merge conflicts. And any saved Justin strings containing that old Justin representation will need to be updated or marked stale.

### Upgrade Considerations

Other than the Compatibility Considerations above, none.

- [x] Update `NEWS.md` for user-facing changes.

### CODE REVIEW

File: packages/marshal/src/marshal-justin.js [hunk 1]
```diff
@@ -445,3 +451,34 @@ const decodeToJustin = (encoding, shouldIndent = false, slots = []) => {
...
+harden(passableAsJustin);
+
+/**
+ * qp for quote passable as a quasi-quoted Justin expression.
+ *
+ * Modelled on `quote` from `assert.js` in `'ses'`. But uses Justin
+ * instead of `bestEffortStringify`. The full name of this would have been
+ * `quotePassable`. But since the `quote` from `assert.js` is always used via
+ * its rename to `q`, we're skipping the rename and just
+ * naming this variable `qp`.
```
Comment (dckc) [inline packages/marshal/src/marshal-justin.js:478]: When somebody looks up `qp`, I think it's less likely that they'll want to know this sort of information about how it's built and more likely that they'll want to know how to use it. An `@example` such as the one in the PR description would be nice.

Please be sure to note that it's not lazy.

I suggest moving these notes about how it's implemented to `//` comments.

File: packages/patterns/test/qp-on-pattern.test.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+// This is a continuation of a test case in marshal-justin.test.js,
+// but is placed here because it uses patterns which don't exist at
+// the `@endo/marshal` layer.
+//
+// This was the original example in a slack thread looking for a more
+// legible way to render patterns. Patterns are Passables, and all
+// Passables can kinda be rendered in Justin. Though Remotables and
+// promises will be rendered only in terms of slot numbers.
+test('qp on a pattern', t => {
```
Comment (dckc) [inline packages/patterns/test/qp-on-pattern.test.js:14]: this is another case where I wish for updoc/doctest, so that the same example can serve as an executable test and documentation.

**Review by dckc — COMMENTED** [review]
some notes on docs

The code looks plausible. I'd have to study it (and the surrounding context) quite a bit more to be confident that it's correct.

File: packages/marshal/src/marshal-justin.js (hunk 1 above)
Comment (erights) [reply packages/marshal/src/marshal-justin.js]: Much better, thanks!

Done

File: packages/marshal/src/marshal-justin.js [hunk 3]
```diff
@@ -445,3 +451,60 @@ const decodeToJustin = (encoding, shouldIndent = false, slots = []) => {
...
+ * @returns {string}
+ */
+export const passableAsJustin = (passable, shouldIndent = true) => {
+  let slotCount = 0;
+  // eslint-disable-next-line no-plusplus
+  const convertValToSlot = val => `s${slotCount++}`;
+  const { toCapData } = makeMarshal(convertValToSlot);
+  const { body, slots } = toCapData(passable);
+  const encoded = JSON.parse(body);
+  return decodeToJustin(encoded, shouldIndent, slots);
```
Comment (dckc) [inline packages/marshal/src/marshal-justin.js:467]: I'm familiar with `decodeToJustin` from its use in the `agoric follow` command:
https://github.com/Agoric/agoric-sdk/blob/777e6a4cefd5c533aab3aeb224f95d5e1edbbd03/packages/agoric-cli/src/follow.js#L76

for example:

```console
$ agoric follow -B https://devnet.agoric.net/network-config :published.agoricNames.brand
[
  [
    "BLD",
    slotToVal("board0566","Alleged: BLD brand"),
  ],
  [
    "IST",
    slotToVal("board0257","Alleged: IST brand"),
  ],
  [
    "Invitation",
    slotToVal("board0074","Alleged: Zoe Invitation brand"),
  ],
...
]
```



In fact, my first reaction to this PR was: we already have `decodeToJustin`; why do we need this?

But now I see that the `makeMarshal` here is simpler than [the makeMarshal in agoric follow](
https://github.com/Agoric/agoric-sdk/blob/777e6a4cefd5c533aab3aeb224f95d5e1edbbd03/packages/agoric-cli/src/follow.js#L19). @michaelfig why does `makeCapDataToQclass` need to keep tables in both directions? Why don't we need that here?

I suspect that there's a test we should add to clarify how these are different.

At a minimum, I'd like to see a test that includes a few remotables.

File: packages/marshal/src/marshal-justin.js [hunk 4]
```diff
@@ -445,3 +451,60 @@ const decodeToJustin = (encoding, shouldIndent = false, slots = []) => {
...
+
+/**
+ * @param {Passable} passable
+ * @param {boolean} [shouldIndent]
+ * @returns {string}
+ */
+export const passableAsJustin = (passable, shouldIndent = true) => {
+  let slotCount = 0;
+  // eslint-disable-next-line no-plusplus
+  const convertValToSlot = val => `s${slotCount++}`;
```
Comment (dckc) [inline packages/marshal/src/marshal-justin.js:463]: consider:

```suggestion
  const convertValToSlot = val => `s${(slotCount += 1)}`;
```

not exactly the same... using `+=` is more like `++slotCount`. But close enough?

Comment (erights) [reply packages/marshal/src/marshal-justin.js:463]: All else being equal, I would prefer that. But I'd to start counting at zero, because that would reflect what the slot indexes would be. And I don't want to initialize to `-1` because that just moves the surprise elsewhere that's not better.

Comment (erights) [reply packages/marshal/src/marshal-justin.js:463]: Resolving. But if you object, please unresolve.

File: packages/marshal/src/marshal-justin.js (hunk 3 above)
Comment (erights) [reply packages/marshal/src/marshal-justin.js:467]: The test case I added to marshal does include two remotables and a promise: https://github.com/endojs/endo/blob/032533a6de48a9e00c21caa1f154856ca89d139b/packages/marshal/test/marshal-justin.test.js#L85

Unlike the other tests, this one is not intended to be a good example, but rather just to probe an edge case.

Comment (erights) [reply packages/marshal/src/marshal-justin.js:467]: If this test satisfies, please resolve this conversation. Thanks.

File: packages/marshal/src/marshal-justin.js (hunk 4 above)
Comment (dckc) [reply packages/marshal/src/marshal-justin.js:463]: I don't feel all that strongly about it, but... what's the justification for this exception to `no-plusplus`? I suggest including it in the comment.

Initializing to `-1` doesn't seem that surprising to me.

File: packages/marshal/src/marshal-justin.js (hunk 3 above)
Comment (michaelfig) [reply packages/marshal/src/marshal-justin.js:467]: `marshal-justin.js`'s `passableAsJustin` accepts a passable and fabricates the slot ids out of thin air (`s${number}`),  where that ID is basically meaningless.  It's unique for every remotable traversed by `toCapData(passable)`, so it can't be used to correlate identity by comparing the slotIDs of two remotables.

The reason for the tables in `capDataToQclass` is to preserve all the semantics of the input capData.  The information described by the `slotIDs` and `ifaces` of the input capData is passed through to the output Justin `makeSlot` calls.

It may very well be that `passableAsJustin` can't recover meaningful `slotIDs`, but @erights, I would encourage you to consider keeping a short-lived WeakMap to reuse an existing `slotID` if the same `remotable` (or even `promise`) appears elsewhere in the input passable.

I don't really remember what the second map was used for.  Humph.

Comment (michaelfig) [reply packages/marshal/src/marshal-justin.js:467]: Oh.  nvm.  Does `@endo/marshal` already track remotable identity already?  I thought that was up to `valToSlot`.

**Review by dckc — APPROVED** [review]
I'm satisfied.

I leave it to you to judge whether to wait for a review from @kriskowal

File: packages/marshal/src/marshal-justin.js (hunk 4 above)
Comment (erights) [reply packages/marshal/src/marshal-justin.js:463]: I merged this PR before adding the requested comment. #2801 would correct.

File: packages/marshal/src/marshal-justin.js [hunk 5]
```diff
@@ -445,3 +451,60 @@ const decodeToJustin = (encoding, shouldIndent = false, slots = []) => {
...
+  const { body, slots } = toCapData(passable);
+  const encoded = JSON.parse(body);
+  return decodeToJustin(encoded, shouldIndent, slots);
+};
+harden(passableAsJustin);
+
+// The example below is the `patt1` test case from `qp-on-pattern.test.js`.
+// Please co-maintain the following doc-comment and that test module.
+/**
+ * `qp` for quote passable as a quasi-quoted Justin expression.
```
Comment (dckc) [inline packages/marshal/src/marshal-justin.js:474]: Now that this landed, I'd expect it to appear in the list of function on https://endojs.github.io/endo/modules/_endo_marshal.html . But I don't see it. Anybody know why not?

Comment (erights) [reply packages/marshal/src/marshal-justin.js:474]: I do not.

**Comment by mhofman** [comment]
> ### Security Considerations
> Unlike `q`, `qp` does not lazily postpone rendering the expression into Justin. `q` can do this easily because it is inside the `'ses'` assert.js module and can use the rights amplification available only there. `qp` depends on `'@endo/marshal'` which is several levels up from `'ses'`. Rather than expose these rights amplification powers (or several other kludges that failed), I just made `qp` non-lazy. Thus there are no security implications.
> 
> ### Scaling Considerations
> Because `qp` is not lazy, simply constructing an error with `Fail` or `makeError(X...)` will pay all the cost of rendering to Justin, whether or not that error is ever used.


My understanding is that `ses`'s `quote` does not attempt to prevent trapping by the quoted object when lazily evaluated.

Looking at `bestEffortStringify` is there a reason it couldn't look for an own `toString` property and invoke that if it exists?

In that case, `qp` could simply create an object that does what this PR does when its `toString` method is invoked, to recover the lazy aspect.

**Comment by erights** [comment]
That's a good idea. I actually tried to do something similar and gave up because, for some reason, it did not occur to me to modify `bestEffortStringify` itself. Without mod, I noticed that it looked for a `Symbol.toString`-named property, and I tried to shoehorn the behavior into a getter for that property. Doing this with a simple non-accessor `toString` seems good.

If I do it, it'll be in PR separate from #2801 . Or, if anyone else wants to try, I'll be happy to review.

**Comment by mhofman** [comment]
> Without mod, I noticed that it looked for a `Symbol.toString`-named property, and I tried to shoehorn the behavior into a getter for that property.

Yeah a `{get [Symbol.toStringTag]() { return qp(payload) }}` would work too, but be a little more hacky.
